subsys/bluetooth/host/l2cap.c

```diff
@@ -1167,25 +1180,45 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
 	}
 
 response:
-	buf = l2cap_create_le_sig_pdu(buf, BT_L2CAP_ECRED_CONN_RSP, ident,
-				      sizeof(*rsp) + (sizeof(scid) * i));
+	if (result != BT_L2CAP_LE_SUCCESS) {
+		if (i == 0) {
+			/* All connections refused */
+			buf = l2cap_create_le_sig_pdu(buf, BT_L2CAP_ECRED_CONN_RSP, ident,
+				      sizeof(*rsp) + sizeof(scid));
 
-	rsp = net_buf_add(buf, sizeof(*rsp));
-	(void)memset(rsp, 0, sizeof(*rsp));
+			rsp = net_buf_add(buf, sizeof(*rsp));
+			(void)memset(rsp, 0, sizeof(*rsp));
+
+			memset(dcid, 0, sizeof(scid));
+			rsp->result = sys_cpu_to_le16(result);
+		} else {
+			buf = l2cap_create_le_sig_pdu(buf, BT_L2CAP_ECRED_CONN_RSP, ident,
+				      sizeof(*rsp) + (sizeof(scid) * i));
 
-	if (result == BT_L2CAP_LE_ERR_UNACCEPT_PARAMS ||
-	    result == BT_L2CAP_LE_ERR_PSM_NOT_SUPP ||
-	    result == BT_L2CAP_LE_ERR_AUTHENTICATION) {
-		memset(dcid, 0, sizeof(scid) * i);
+			rsp = net_buf_add(buf, sizeof(*rsp));
+			(void)memset(rsp, 0, sizeof(*rsp));
+			memset(dcid, 0, sizeof(scid) * i);
+			rsp->result = sys_cpu_to_le16(result);
+		}
 	} else if (ch) {
+		buf = l2cap_create_le_sig_pdu(buf, BT_L2CAP_ECRED_CONN_RSP, ident,
+				      sizeof(*rsp) + (sizeof(scid) * i));
+
+		rsp = net_buf_add(buf, sizeof(*rsp));
+		(void)memset(rsp, 0, sizeof(*rsp));
 		rsp->mps = sys_cpu_to_le16(ch->rx.mps);
 		rsp->mtu = sys_cpu_to_le16(ch->rx.mtu);
 		rsp->credits = sys_cpu_to_le16(ch->rx.init_credits);
+		rsp->result = sys_cpu_to_le16(result);
 	}
 
-	net_buf_add_mem(buf, dcid, sizeof(scid) * i);
+	if (i == 0) {
+		/* All connections refused */
+		net_buf_add_mem(buf, dcid, sizeof(scid));
```

Comment:
If all connections are refused and we send only one entry then I think this would be easiest to handle by just having
`net_buf_add_le16(buf, 0);` and then get rid of the memset above of dcid. It just makes it harder to read the code, and memseting of dcid just to add a zero here is not very straightforward.